### PR TITLE
pm: device power management improvements

### PR DESCRIPTION
**Pull request by gmarull**
Summary:

- Remove usage of local states, the device PM subsystem already holds device PM state
- The "ctrl_command" is removed, since the PM control callback will now always be called to set the state, not to retrieve it.
- The state is passed as a value, since it does not need to be written by the PM control callback
- The callback is only called for a state change, thus simplifying driver operation (no need for state change checks)
- Converged _low power_ states to `PM_DEVICE_STATE_SUSPEND`. This makes interoperability between system and device PM easier. The low power state is still kept, but its future is unclear and needs some discussion.
- Device PM control function return codes are now documented.
- Multiple implementation cleanups.
- Introduction of the _actions_ concept. Devices will now receive actions instead of states, for example, a device will be told to "suspend" or to "resume".

### CODE REVIEW

**Review by carlescufi — APPROVED** [review]
+1 to the approach after discussion with @gmarull

File: drivers/interrupt_controller/intc_ioapic.c
Comment (ceolin) [inline drivers/interrupt_controller/intc_ioapic.c:327]: Original code is handling `PM_DEVICE_STATE_LOW_POWER`. Here you are not doing it and it will return `-ENOTSUP` instead of `0`.

Comment (gmarull) [reply drivers/interrupt_controller/intc_ioapic.c:327]: It's not handling it really, it does nothing. 
```
		case PM_DEVICE_STATE_LOW_POWER:
			break;
```

I can leave as it is and fix this problem later. I don't think PM subsys is handling `-ENOTSUP` correctly as of today.

Comment (gmarull) [reply drivers/interrupt_controller/intc_ioapic.c:329]: added the case back, will fix these sort of issues later.

File: drivers/spi/spi_nrfx_spi.c
Comment (ceolin) [inline drivers/spi/spi_nrfx_spi.c:297]: you are unconditionally calling `nrfx_spi_uninit` while the original code just call it if the current state is `ACTIVE`.

File: samples/subsys/pm/device_pm/src/dummy_driver.c
Comment (ceolin) [inline samples/subsys/pm/device_pm/src/dummy_driver.c:36]: `(void)pm_device_state_get(dev, &state);`

**Review by ceolin — CHANGES_REQUESTED** [review]

File: drivers/serial/uart_cc13xx_cc26xx.c
Comment (gmarull) [reply drivers/serial/uart_cc13xx_cc26xx.c:432]: there's an upcoming patch cleaning up driver implementations (https://github.com/zephyrproject-rtos/zephyr/pull/36125/commits/2aecc31fbbed9b98cb2d0af918a66b3d43c000c1), it's out of the scope of this PR.

File: drivers/serial/uart_npcx.c
Comment (gmarull) [reply drivers/serial/uart_npcx.c:463]: ref. #36125

File: drivers/spi/spi_nrfx_spi.c
Comment (gmarull) [reply drivers/spi/spi_nrfx_spi.c:297]: this is precisely the purpose of this PR: remove this redundant logic. There is no need to do this check, since the device will be in active state if it reaches this point.

File: drivers/spi/spi_nrfx_spim.c
Comment (gmarull) [reply drivers/spi/spi_nrfx_spim.c:344]: https://github.com/zephyrproject-rtos/zephyr/pull/36759#discussion_r672528317

File: subsys/pm/device.c
Comment (gmarull) [reply subsys/pm/device.c:156]: they all end up running the same check, since they don't have transitional states:
```
		if (dev->pm->state == state) {
			return -EALREADY;
		}
```

File: samples/subsys/pm/device_pm/src/dummy_driver.c
Comment (gmarull) [reply samples/subsys/pm/device_pm/src/dummy_driver.c]: fixed

File: drivers/sensor/bmp388/bmp388_trigger.c
Comment (gmarull) [reply drivers/sensor/bmp388/bmp388_trigger.c]: fixed

File: drivers/serial/uart_cc13xx_cc26xx.c
Comment (ceolin) [reply drivers/serial/uart_cc13xx_cc26xx.c:432]: That is not out of this scope, since it is this patch that is causing this function no longer be needed. It is definitely out of scope, but I can live with it being addressed in another pr.

File: subsys/pm/device.c
Comment (ceolin) [reply subsys/pm/device.c:156]: gotcha.

File: drivers/spi/spi_nrfx_spi.c
Comment (ceolin) [reply drivers/spi/spi_nrfx_spi.c:297]: One calls `pm_device_state_set(dev, PM_DEVICE_STATE_LOW_POWER);` if the device is ACTIVE it will result in this code calling `nrfx_spi_uninit(&config->spi);`. Now the device is `PM_DEVICE_STATE_LOW_POWER`.

Now, if we do: `pm_device_state_set(dev, PM_DEVICE_STATE_OFF);` this function will end up calling  `nrfx_spim_uninit(&config->spim);` again right ? While in the original implementation this was only called if the device was ACTIVE.

File: drivers/spi/spi_nrfx_spim.c
Comment (ceolin) [reply drivers/spi/spi_nrfx_spim.c:344]: https://github.com/zephyrproject-rtos/zephyr/pull/36759#discussion_r672548573

File: subsys/pm/device.c
Comment (ceolin) [reply subsys/pm/device.c:156]: that is correct.

File: drivers/spi/spi_nrfx_spi.c
Comment (gmarull) [reply drivers/spi/spi_nrfx_spi.c:297]: Indeed, this is a PM framework problem because we have poorly defined states and transitions, leading to this sort of issues. In a first place, `pm_device_state_set` needs to be removed from the public API space as we all agree it is not the proper way to handle device PM. Second, most (if not all) devices have this sort of fall-through to handle all "low power states", which is wrong. The possible state transitions should be limited, e.g. as Windows does https://docs.microsoft.com/en-us/windows-hardware/drivers/kernel/device-power-states

I have just opened an issue detailing all these problems https://github.com/zephyrproject-rtos/zephyr/issues/37056

#36125 goes further simplifying most of these problems, plus introducing the actions concept, which helps to prevent this sort of problems.

Comment (ceolin) [reply drivers/spi/spi_nrfx_spi.c:297]: Agree, but we should stop break everything thing while we don't have proper solutions. I like the idea of having defined transitions so devices don't need to check it by themselves. But even considering what is done by windows in https://docs.microsoft.com/en-us/windows-hardware/drivers/kernel/device-power-states, it is possible to go from D3Hot to D3Cold but it is not defined that all devices should support it.

So lets do an exercise, imagine this devices maps `PM_DEVICE_STATE_SUSPEND` to `D3Hot` and `PM_DEVICE_STATE_OFF` to `D3Cold`. In this scenario the transition between `PM_DEVICE_STATE_SUSPEND` to `PM_DEVICE_STATE_OFF` is allowed from the subsystem perspective, but this driver does not support it. It just change the state if the current state is `ACTIVE`.  With your patch this transition will happen. 

That is wrong ? I don't know, but I don't like change device drivers behavior in this massive changes. I vote to keep their behavior and after this being merged address it in another pr, that is smaller and will get attention from the proper maintainer. We *can't* assume we know everything about all devices.

Comment (ceolin) [reply drivers/spi/spi_nrfx_spi.c:297]: Btw, that is my only concern in this pr right now, everything else is great for me.

Comment (gmarull) [reply drivers/spi/spi_nrfx_spi.c:297]: I think the problem is that most drivers started following the pattern below because there is no clear guidance on what low power states mean as of today, so everything that sounded like "low power" was put into the same basket:

```c
	case PM_DEVICE_STATE_LOW_POWER:
	case PM_DEVICE_STATE_SUSPEND:
	case PM_DEVICE_STATE_OFF:
```

This design pattern has some problems. For example, suspend and off should not perform the same action since one will keep context while the other won't. There are actually many other problems that this PR is not fixing, see e.g.

```c
		if (*state == PM_DEVICE_STATE_ACTIVE) {
			ret = st7735r_exit_sleep(data);
			if (ret < 0) {
				return ret;
			}
			data->pm_state = PM_DEVICE_STATE_ACTIVE;
		} else {
			ret = st7735r_enter_sleep(data);
			if (ret < 0) {
				return ret;
			}
			data->pm_state = PM_DEVICE_STATE_LOW_POWER;
		}
```

In this case things are even worse, since any state `!= PM_DEVICE_STATE_ACTIVE` will call the sleep code and leave the device in `PM_DEVICE_STATE_LOW_POWER`, even if another state was requested.

Note that #36125 RFC fixes most (if not all) these problems in later patches by removing these bad patterns. But still, I think that until #37056 is solved, we won't have a final/proper solution.

Comment (gmarull) [reply drivers/spi/spi_nrfx_spi.c:297]: For reference, the final tentative code proposed in the RFC for the mentioned drivers:

https://github.com/zephyrproject-rtos/zephyr/blob/cebb50a54e375f1e5c7af96ae38d1602a70c2531/drivers/spi/spi_nrfx_spi.c#L280-L303

https://github.com/zephyrproject-rtos/zephyr/blob/cebb50a54e375f1e5c7af96ae38d1602a70c2531/drivers/display/display_st7735r.c#L510-L529

**Review by nashif — CHANGES_REQUESTED** [review]
Changing the control API requires further discussion, the cleanup in general is ok, but it does not provide an alternate long-term solution.

File: drivers/display/display_st7735r.c
Comment (gmarull) [reply drivers/display/display_st7735r.c:515]: This is why actions (or call them commands) are introduced in the #36125 RFC.

Comment (gmarull) [reply drivers/display/display_st7735r.c:515]: In the end, all these changes simplify device PM and at the same time align a bit more with what Linux does. I think these changes do not fix all the problems of device PM, but I think they move to the right direction. If one looks at device PM today, it's really in a sad state (very few drivers implement it and most have bugs). #36125 or #37056 provide enough evidence to prove that. I haven't seen any alternatives or roadmap to improve the large amount of PM issues we have today.

Comment (nashif) [reply drivers/display/display_st7735r.c:515]: > This is why actions (or call them commands) are introduced in the #36125 RFC.

ok, that was exactly on my mind and makes sense. You need better isolate changes in separate PRs. This PR on its own does not make sense without the move to actions. The other Pr is full of conflicts and I am not sure anymore which is which, every PR starts with "simplify"... :)

Comment (gmarull) [reply drivers/display/display_st7735r.c:515]: Just rebased #36125 so that it's easier to browse.

**Comment by gmarull** [comment]
Before doing any more changes, I'd like to have a clear roadmap on where we want to go, otherwise every PR will take ages to merge and as a result we'll keep having a device PM subsys that is far from solving the PM problem. I think that if the goal is clear, and the current PM subsys status is well understood/recognized, things will be easier. I have the feeling that device PM in particular is _sold_ as ready when it is clearly not, and, as far as I know, I'm not the first person to mention that.

Let's take a couple of examples to illustrate my argument:

- The `ctrl_command` argument was, in theory, added to support arbitrary commands. The reality, however, is that it has only been used to set/get the device state (with a command name `PM_DEVICE_STATE_GET/SET`, same prefix as actual device states). I think everybody agrees that keeping local states is, in general, a bad idea. Removing the "get", we're left with a single command to set the state. So in reality, the callback is de-facto being used with an _action_, that is, we combine a "set" command with a "state" argument. An example to make it clear: `"set + active" == "resume"`. The `void *context` was replaced in the last release in favor of a `uint32_t *state` and now with `enum pm_device_state *state`, meaning a generic command can hardly be implemented now. So the reason to keep `ctrl_command` doesn't hold anymore. What this PR does in the end is to continue the same design pattern (use states as actions), just removing a redundancy (the ctrl_command), so I don't see what is wrong with this approach as an intermediate step to a proper solution as suggested in #36125.
- The runtime API is another example of something that has existed, but nobody (except a single driver) uses it. It lacks proper documentation and usage guidelines, it doesn't handle device dependencies, etc. Let's talk about a problem I have already mentioned with this API. It basically puts devices into `ACTIVE` or `SUSPEND` states. But what about `LOW_POWER`? Let's see: according to the specs, `LOW_POWER` should be used when context is not lost. In many devices this holds for low power states, so a driver for such device will pick `LOW_POWER`. But then, how can runtime PM be implemented? What happened in reality: most drivers handle all the "low power states" the same way, which is wrong and leads to inconsistent behavior (it's incompatible to implement low power/suspend/off the same way). This is a common pattern in the PM subsystem: there are many APIs that have been created without really analyzing if they can solve a real problem.

My feeling is that the PM subsystem was designed in a rush, with a quick implementation that likely solved immediate problems, but it is now showing its limitations and requires some fundamental changes. We're not talking about some fixes, API extensions, etc. but a refactor/re-write in some areas. This is difficult, painful and likely means a period of some instability, but it is sometimes necessary to make progress. And let's be honest: we'll not either get it 100% right this time.

**Review by ceolin — APPROVED** [review]

**Comment by ceolin** [comment]
The cleanup is fine, that is definitely not the final version of this control API, but lets move forward.

**Comment by gmarull** [comment]
@ceolin @nashif I've added all the changes up to initial support of actions.

**Review by nashif — APPROVED** [review]

**Pull request merged**
